multiple RPM-related packages: some TLC

Kangie opened the pull request:
This PR updates createrepo_c to 0.20.1-r1 (now with working tests!) and adds a number of outstanding dependencies that make createrepo_c a compelling alternative to the legacy createrepo script.

Due to the new dependencies I have dropped `~x86` from `KEYWORDS` as I'm not currently in a position to test them all.

Kangie commented:
@thesamesam 

Few QA issues here:

- libmodulemd is throwing '... python modules that are not byte compiled' - It looks to me like [cause 3](https://projects.gentoo.org/python/guide/qawarn.html#compiled-bytecode-related-warnings) from the docs - am I correct in that assessment?
- ~drpm throws `warning: incompatible implicit declaration of built-in function` x4 - Florian has fixed this upstream but we don't have a new package yet. I'm willing to bump to `_p<date>` or pull in florian's patch~ - I'll do that now!

thesamesam commented on app-arch/createrepo_c/createrepo_c-0.20.1-r1.ebuild:
```diff
@@ -0,0 +1,80 @@
+	drpm? ( app-arch/drpm )"
+
+RDEPEND="${DEPEND}
+	app-arch/xz-utils
```
Just checking: it doesn't use liblzma, only calls `xz` at runtime (or it `dlopen()`s liblzma?)

Kangie replied:
I will have to look into this more. My initial assumption was wrong, it is clearly using `liblzma` as an include!

thesamesam commented on app-arch/createrepo_c/createrepo_c-0.20.1-r1.ebuild:
```diff
@@ -0,0 +1,80 @@
+	cmake_src_configure
+}
+
+distutils_enable_tests unittest
```
Put this above src_prepare

thesamesam commented on app-arch/drpm/files/drpm-0.5.1-c99.patch:
```diff
@@ -0,0 +1,24 @@
+From 0bde9b7711a2fa8b811830cadfc979365f9362c6 Mon Sep 17 00:00:00 2001
```
Link to upstream commit. But good spot ;)

thesamesam commented on app-arch/zchunk/zchunk-1.2.3.ebuild:
```diff
@@ -0,0 +1,28 @@
+
+LICENSE="BSD-2"
+SLOT="0"
+IUSE=""
```
drop

thesamesam commented on sys-libs/libmodulemd/libmodulemd-2.14.0.ebuild:
```diff
@@ -0,0 +1,58 @@
+		$(meson_use gtk-doc with_docs)
+		$(meson_use test test_installed_lib)
+	)
+	# There isn't an option to skip valgrind tests at this point, and the whole test suite takes 7s.
```
This is _tolerable_ for now if you must, but I prefer to not run Valgrind at all for tests, because it's not portable to other arches.

thesamesam commented on sys-libs/libmodulemd/metadata.xml:
```diff
@@ -0,0 +1,15 @@
+			<name>Proxy Maintainers</name>
+	</maintainer>
+	<maintainer type="person" proxied="yes">
+			<email>Matt.Jolly@footclan.ninja</email>
```
Put yourself first

thesamesam commented on sys-libs/libmodulemd/metadata.xml:
```diff
@@ -0,0 +1,15 @@
+			<name>Matt Jolly</name>
+	</maintainer>
+	<upstream>
+		<remote-id type="github">rpm-software-management/createrepo_c</remote-id>
```
Is this right?

Kangie replied:
Guess which of these was my template ;)

thesamesam commented on app-arch/createrepo_c/metadata.xml:
```diff
@@ -9,6 +9,10 @@
 	</maintainer>
+	<use>
+		<flag name="drpm">Enable delta RPM support via <pkg>app-arch/drpm</pkg></flag>
+		<flag name="legacy">Build with support for legacy weakdeps and hashes (not recommended!)</flag>
```
I assume you actually needed this in the wild, hence had to add an option for it? (Otherwise I'd say we just shouldn't do it at all)

Kangie replied:
tbh `drpm` could probably be unconditionally enabled - I can't see that enabling having delta RPM support will break anything.

The `legacy` flag enables both legacy suse/mandriva weakdeps support (as opposed to the 'modern' weakdeps support, I'm assuming - basically the equivalent of ebuild optfeature) and the use of SHA1 & MD5 for digests. I can see a use case where using the modern createrepo_c implementation to manage legacy RPM repositories is desirable - there are some edge cases in the legacy script for basic functionality. I think that I'd prefer that it remain toggleable for now, eventually legacy distro releases will age out.

If you have a strong preference for getting rid of the 'legacy' flag I'm happy remove it; I don't think that *my* use case is quite old enough to need it, it's just within throwing distance!

thesamesam commented:
I need to check your remaining questions, remind me in a day if I forget.

Kangie commented:
Few updates here for posterity:

- `libmodulemd` has a new patch. I'm not sure that upstream will apply it in the current form, but it only "breaks" EPEL7 packaging because their meson is too old for features this package is already using; for us it removes a bunch of concerning (and legitimate) warnings from build output.
- I've elected to drop the `python` portions of `createrepo_c` - it doesn't make too much sense for Gentoo (though we *do* lose out on python bindings). My understanding is that the build will result in either python (wheel) or C output. There's not a middle ground where you get python bindings *and* C libs to install to the system - at least not while using distutils.
  + A maintainer who needs the python bindings can dig into this further :)
- I'm finalising an upstream patch for `createrepo_c` to make it more portable - currently createrepo_c relies on coreutils `cp` binary with GNU extensions (--preserve/-p). I'll get that attached to this PR and amend for any other feedback before we merge.

jkroonza commented on app-arch/createrepo_c/createrepo_c-0.20.1-r1.ebuild:
```diff
@@ -0,0 +1,79 @@
+	sys-libs/zlib:=
+"
+
+RDEPEND="${DEPEND}"
```
Dropping lzma?

Kangie replied:
`app-arch/xz-utils` is the upstream dependency and provides liblzma/lzma.h

jkroonza commented on app-arch/createrepo_c/createrepo_c-0.20.1-r1.ebuild:
```diff
@@ -0,0 +1,79 @@
+RDEPEND="${DEPEND}"
+
+BDEPEND="
+	${DEPEND}
```
This seems unlikely.  This should be for tooling required to build, so does the build system for example invoke the curl CLI command line tool?  And openssl ...?  Etc.

Kangie replied:
Fixed!

jkroonza commented on app-arch/createrepo_c/createrepo_c-0.20.1-r1.ebuild:
```diff
@@ -0,0 +1,79 @@
+
+src_prepare() {
+	cmake_src_prepare
+}
```
I do not believe src_prepare override here should be necessary?  Doesn't cmake_src_prepare get called by default?

Kangie replied:
Good catch; shrapnel from when I was experimenting with python bindings and tests.

FWIW I was able to get python building and all tests passing using the distutils eclass, however it seems to be a mutually exclusive build - either C or Python stuff gets installed at the end. I'm going to chat to upstream and find out if I'm missing something, because it would be nice to have some importable python modules for scripting.

That'll be a `-r2` though.

jkroonza replied:
ACK, yes, would be nice, we're just invoking the CLI tool from our end, which suits our needs perfectly.

jkroonza commented on app-arch/createrepo_c/createrepo_c-0.20.1-r1.ebuild:
```diff
@@ -0,0 +1,79 @@
+src_configure() {
+	local mycmakeargs=(
+		-DENABLE_DRPM=ON
+		# As best I can tell, this enables distribution as a wheel. No need for this on gentoo!
```
What is a wheel in the context?

Kangie replied:
A wheel is a binary distribution format of a python package as (currently) defined [here](https://packaging.python.org/en/latest/specifications/binary-distribution-format/#binary-distribution-format)

jkroonza replied:
Right.  Python stuff I don't particularly care too much about, and which may well be useful to others, but not currently myself.  I don't write much python code.

jkroonza commented on app-arch/createrepo_c/createrepo_c-0.20.1-r1.ebuild:
```diff
@@ -0,0 +1,79 @@
+
+src_configure() {
+	local mycmakeargs=(
+		-DENABLE_DRPM=ON
```
This changes the value - why?

Kangie replied:
Because we have the drpm dep in gentoo now, there's no reason for it to be toggleable.

jkroonza commented on app-arch/createrepo_c/createrepo_c-0.20.1-r1.ebuild:
```diff
@@ -0,0 +1,79 @@
+		# Upstream enables some 'Legacy' stuff by default, let's put that behind a USE flag
+		-DENABLE_LEGACY_WEAKDEPS=$(usex legacy ON OFF)
+		-DWITH_LEGACY_HASHES=$(usex legacy ON OFF)
+		-DWITH_ZCHUNK=ON
```
Inversion - why?

Kangie replied:
Because we have the zchunk dep in gentoo now, there's no reason for it to be toggleable.

jkroonza replied:
we do?  I can find bchunk, but I'm failing to find zchunk?  Probably just looking right over it.

jkroonza commented on app-arch/createrepo_c/createrepo_c-0.20.1-r1.ebuild:
```diff
@@ -0,0 +1,79 @@
+		-DENABLE_LEGACY_WEAKDEPS=$(usex legacy ON OFF)
+		-DWITH_LEGACY_HASHES=$(usex legacy ON OFF)
+		-DWITH_ZCHUNK=ON
+		-DWITH_LIBMODULEMD=ON
```
Inversion - why?

Kangie replied:
Because we have the libmodulemd dep in gentoo now, there's no reason for it to be toggleable.

jkroonza reviewed (commented):
Thanks for your contributions.  Really appreciate, just need some clarity.

Looks like primarily you're adding tests here, which is a good thing.

Not sure I'm in favour of a -9999 ebuild ... but sure if there is motivation for that.

Parts of the use-gio-instead-of-cp patch us style cleanups (removal of spurious newlines).  Not sure these should be included as it increases the risks of future conflicts.

I know that md5 and sha1 is considered weak, but I believe 'legacy' feels like a bit of a wrong name, and we do still have customers with VERY old systems for which we do need to build RPMs that can install on those systems, so for the time being can we please just keep those unconditionally.  USE flag is probably ok, but 'weak-hashes' is then probably a better name than 'legacy'.

Kangie commented:
> Not sure I'm in favour of a -9999 ebuild ... but sure if there is motivation for that.

It provides a template for any future version bumps, and if you have a reason to build from source quickly (say to check if the bug you've just uncovered has been resolved) it's trivial to unmask and test.

> Parts of the use-gio-instead-of-cp patch us style cleanups (removal of spurious newlines). Not sure these should be included as it increases the risks of future conflicts.

It's the PR as a patch as it was submitted upstream. I could amend it, but the intent is to get this merged so that we can drop this patch for the next release. Besides, I need it merged upstream so that it works on the distros with `cp` that doesn't like `--preserve` ::wink::

> I know that md5 and sha1 is considered weak,but I believe 'legacy' feels like a bit of a wrong name, ... USE flag is probably ok, but 'weak-hashes' is then probably a better name than 'legacy'.

Legacy is what upstream calls it!

I also note that the 'legacy' use flag covers both legacy-style (suse, mandriva) weakdeps support and the legacy hashes as kind of a 'if you want old stuff you can have old stuff' - any thoughts on that?

If you have suggestions as to other terminology I'm interested, the best I can come up with otherwise is "crufty-old-deprecated-rpms" ::wink::, however I don't see much value into splitting 'legacy configure options' into two USE flags.

>  we do still have customers with VERY old systems for which we do need to build RPMs that can install on those systems, so for the time being can we please just keep those unconditionally.

I'm willing to consider unconditionally enabling these options, however there's also a point to be made about being secure 'by default'; I'd prefer that we enable insecure/legacy/weak hashes explicitly.

My thoughts are: If you need to maintain repos for legacy systems it's fair to enable the `legacy` USE and explicitly acknowledge that you're using enabling hashes vulnerable to collisions.

Kangie commented:
@jkroonza After discussion with Sam the other day I'm not looking to merge this until upstream have had a chance to comment on the gio patch, but it's passing their CI/CD; I'm pretty confident that it'll go ahead as-is.

jkroonza commented:
@Kangie if you need this, I have zero objection to merging it now on our side now.  Given the explanations above, I'm happy.  You're also welcome to add yourself as co-maintainer to createrepo_c - I've got one specific use-case for this, so this doesn't end up receiving a lot of TLC from myself, you plainly care more in-depth about this than myself.  Appreciate the TLC you gave this, thank you.

Kangie commented:
Should be good to go. Updated the patch to the one that was accepted upstream.

I've dropped the `doc` USE:

- the BDEPEND was wrong - should actually be `doc? ( app-doc/doxygen[dot] )`
- It requires an additional build target `doc` which is not in the previous ebuild
- The `doc` target requires the python components to be build and I'm not willing to untangle this right now.

I've had another play with Python builds - this is an attempt to do both, I still feel like it's an either/or, maybe?

Required Ebuild changes: 

```bash
DISTUTILS_USE_PEP517=setuptools
DISTUTILS_SINGLE_IMPL=1
PYTHON_COMPAT=( python3_{10..11} )

inherit cmake distutils-r1

IUSE="doc legacy python test"
REQUIRED_USE="doc? ( python ) 

BDEPEND="
	doc? ( app-doc/doxygen[dot] )
	python? ( dev-python/scikit-build[${PYTHON_SINGLE_USEDEP}] )
"

src_prepare() {
	# we don't want to run python src_prepare stuff or we'll patch twice!
	cmake_src_prepare
}

src_configure() {
...
# This should work but it doesn't override the built-in cmake_args array in setup.py!
	if use python; then
	    cat >> setup.cfg <<-EOF
        [cmake_args]
		-DENABLE_DRPM=ON
		-DENABLE_LEGACY_WEAKDEPS=$(usex legacy ON OFF)
		-DWITH_LEGACY_HASHES=$(usex legacy ON OFF)
		-DENABLE_PYTHON=ON
		-DWITH_LIBMODULEMD=ON
		-DWITH_ZCHUNK=ON
		#...
		EOF
		esetup.py saveopts
	fi
}

src_compile() {
	cmake_src_compile
	# Tests have a magic target!
	use test && cmake_src_compile tests
	use python && distutils-r1_python_compile
}

...

python_install() {
	distutils-r1_python_install

	# ensure data files for tests are getting installed too
	python_moduleinto collada/tests/
	python_domodule collada/tests/data
}


src_install() {
	python_install
	cmake_src_install
}
```

I *can* get CMake to spit out python modules in addition to its C libs, and install them via `cmake_src_install` however I'm sure that _not_ using distutils for this is a bad idea. Something about wanting to have the correct python impl selected. :)

Anyway, I don't care enough about API documentation to dig into this further, and I'm not sure that adding a python wrapper for something we can compile and install anyway is worthwhile - this is Gentoo.

I feel like if the configure issues can be resolved this could easily be turned into an either/or ebuild with USE=python selecting 'I want a wheel' without too much effort or maintenance burden, and I can see some possible utility if you want to be able to hook into python for some scripting. I'm just not sure how to make the two builds not clobber each other. Perhaps always have Python off in CMake and let the distutils build do its own CMake builds - no guarantee that it won't break something by clobbering C output though.

@thesamesam, I'd be willing to put a little more time into this (in a new ebuild release) if you had some thoughts on gluing this all together... or we can just split the python portion off into its own dev-python/createrepo_c ebuild and deal with it that way.

Kangie commented:
and fixed a few nits :)

jkroonza reviewed (approved):
@thesamesam I'm happy with the explanations given and have no objections.  Not specifically tested my use-cases but doing so is hard due the wide range of systems we bump into.

Kangie commented:
@thesamesam ping!

thesamesam commented:
I'm not particularly worried about the Python side for now. Thanks!
